service/operator/operatorkit.go

```diff
@@ -0,0 +1,146 @@
+	// regardless creation or deletion events.
+	GetEmptyState() interface{}
+	// GetCreateState receives the custom object observed during TPR watches. It
+	// also receives the current state as provided by GetCurrentState and the
+	// desired state as provided by GetDesiredState. GetCreateState analyses the
+	// current and desired state and returns the state intended to be created by
+	// ProcessCreateState.
+	GetCreateState(obj, currentState, desiredState interface{}) (interface{}, error)
+	// GetDeleteState receives the custom object observed during TPR watches. It
+	// also receives the current state as provided by GetCurrentState and the
+	// desired state as provided by GetDesiredState. GetDeleteState analyses the
+	// current and desired state and returns the state intended to be deleted by
+	// ProcessDeleteState.
+	GetDeleteState(obj, currentState, desiredState interface{}) (interface{}, error)
+	// ProcessCreateState receives the custom object observed during TPR watches.
+	// It also receives the state intended to be created as provided by
+	// GetCreateState. ProcessCreateState only has to create resources based on
+	// its provided input. All other reconciliation logic and state transformation
+	// is already done at this point of the reconciliation loop.
+	ProcessCreateState(obj, createState interface{}) error
+	// ProcessDeleteState receives the custom object observed during TPR watches.
+	// It also receives the state intended to be deleted as provided by
+	// GetDeleteState. ProcessDeleteState only has to delete resources based on
+	// its provided input. All other reconciliation logic and state transformation
+	// is already done at this point of the reconciliation loop.
+	ProcessDeleteState(obj, deleteState interface{}) error
+}
+
+// ProcessCreate is a drop-in for an informer's AddFunc. It receives the custom
+// object observed during TPR watches and anything that implements Operator.
+// ProcessCreate takes care about all necessary reconciliation logic for create
+// events.
+//
+//     func addFunc(obj interface{}) {
+//         err := ProcessCreate(obj, operator)
+//         if err != nil {
+//             // error handling here
+//         }
+//     }
+//
+//     newResourceEventHandler := &cache.ResourceEventHandlerFuncs{
+//         AddFunc:    addFunc,
+//     }
+//
+func ProcessCreate(obj interface{}, operator Operator) error {
+	currentState, err := operator.GetCurrentState(obj)
+	if err != nil {
+		return microerror.MaskAny(err)
+	}
+
+	desiredState, err := operator.GetDesiredState(obj)
+	if err != nil {
+		return microerror.MaskAny(err)
+	}
+
+	createState, err := operator.GetCreateState(obj, currentState, desiredState)
+	if err != nil {
+		return microerror.MaskAny(err)
+	}
+
+	err = operator.ProcessCreateState(obj, createState)
+	if err != nil {
+		return microerror.MaskAny(err)
+	}
+
+	deleteState, err := operator.GetDeleteState(obj, currentState, desiredState)
```

Comment:
> The create events provided by the informer is only a event notification. It is not a specification to only be allowed to create state.

That sounds surprising. I would rather structure it so that event handlers only perform actions that match the event type. It feels like a cleaner structure.

But I guess reconciling everything on every event, no matter the type, might make the loop tighter. Not sure.